packages/mock-block-dock/src/datastore/use-mock-datastore.ts

```diff
@@ -0,0 +1,618 @@
+import {
+  EmbedderGraphMessageCallbacks,
+  Entity,
+  Subgraph,
+} from "@blockprotocol/graph";
+import { getEntity as getEntityFromSubgraph } from "@blockprotocol/graph/stdlib";
+import { useCallback, useState } from "react";
+import { v4 as uuid } from "uuid";
+
+import { aggregateEntities as aggregateEntitiesImpl } from "./hook-implementations/entity/aggregate-entities";
+import { getEntity as getEntityImpl } from "./hook-implementations/entity/get-entity";
+import { mockDataToSubgraph } from "./mock-data-to-subgraph";
+import { addEntitiesToSubgraph } from "./mutate-subgraph";
+
+export type MockData = {
+  entities: Entity[];
+  // linkedAggregationDefinitions: LinkedAggregationDefinition[];
+};
+
+export type MockDatastore = {
+  graph: Subgraph;
+  graphServiceCallbacks: Required<EmbedderGraphMessageCallbacks>;
+};
+
+const readonlyErrorReturn: {
+  errors: [{ code: "FORBIDDEN"; message: string }];
+} = {
+  errors: [
+    {
+      code: "FORBIDDEN",
+      message: "Operation can't be carried out in read-only mode",
+    },
+  ],
+};
+
+export const useMockDatastore = (
+  initialData: MockData = {
+    entities: [],
+    // linkedAggregationDefinitions: [],
+  },
+  readonly?: boolean,
+): MockDatastore => {
+  const [graph, setGraph] = useState(mockDataToSubgraph(initialData));
```

Comment:
I used `useDefaultState` here and it cost me about an hour because the state would seemingly get reset on every render. I don't really get the point of the hook (`useDefaultState`)

Reply:
The point of `useDefaultState` is to initialise a local state from the value passed to `useDefaultState`, and then re-initialiaze it if that value changes on susbequent renders (so that callers of this function can affect the local state by passing a new value for the 'default state' argument).

e.g. if I have MBD loaded in some app, and set `initialData` through its props, that initial datastore will get initialized in this local state. If I then want to change the initial datastore mid-session, I cannot do that via MBD props, because this normal `useState` is going to ignore the value passed to it after the first initialization.

The reason it kept re-rendering is because `mockDataToSubgraph(initialData)` returns a new value on every render (it creates a new object in memory), so as far as `useDefaultState` is concerned, it's getting a new value to reinitialize with on every render.

The fix would be to memoize the return of `mockDataToSubgraph`, e.g.

```ts
export const useMockDataToSubgraph = (mockData: MockData) => {
  const subgraph = useMemo(() => {
    const { entities } = mockData;

    const subgraph: Subgraph = {
      roots: [],
      vertices: {},
      edges: {},
      depths: {
        hasLeftEntity: { incoming: 255, outgoing: 255 },
        hasRightEntity: { incoming: 255, outgoing: 255 },
      },
    };

    addEntitiesToSubgraph(subgraph, entities);
  }, [mockData]);

  return subgraph;
};
```

Reply:
Ah that makes sense, thank you for the great explanation!

Glad I opened a comment about this because I sensed something was off.

Fixed in [Turn mockDataToSubgraph into a hook](https://github.com/blockprotocol/blockprotocol/pull/830/commits/221b45e3773e968578f92ed2c6ed499f4af4391f)